datatypes/src/raster/operations/checked_scaling.rs

```diff
@@ -0,0 +1,395 @@
+use crate::raster::{
+    EmptyGrid, GridOrEmpty, GridOrEmpty2D, GridSize, MapElements, MaskedGrid, RasterTile2D,
+};
+use num_traits::{CheckedAdd, CheckedDiv, CheckedMul, CheckedSub};
+use std::ops::{Add, Div, Mul, Sub};
+
+/// scales the value with `(value - offset_by) / scale_with`.
+#[inline]
+#[allow(clippy::unnecessary_wraps)]
+fn scale<T>(value: T, scale_with: T, offset_by: T) -> Option<T>
+where
+    T: Copy + 'static + Sub<Output = T> + Div<Output = T>,
+{
+    Some((value - offset_by) / scale_with)
+}
+
+/// unscales the value with `value * scale_with + offset_by`.
+#[inline]
+#[allow(clippy::unnecessary_wraps)]
+fn unscale<T>(value: T, scale_with: T, offset_by: T) -> Option<T>
+where
+    T: Copy + 'static + Add<Output = T> + Mul<Output = T>,
+{
+    Some(value * scale_with + offset_by)
+}
+/// scales the value with `(value - offset_by) / scale_with`. Overflows produce `None`.
+#[inline]
+fn scale_checked<T>(value: T, scale_with: T, offset_by: T) -> Option<T>
+where
+    T: Copy + 'static + CheckedSub<Output = T> + CheckedDiv<Output = T>,
+{
+    value
+        .checked_sub(&offset_by)
+        .and_then(|f| f.checked_div(&scale_with))
+}
+
+/// unscales the value with ``value * scale_with + offset_by``. Overflows produce `None`.
+#[inline]
+fn unscale_checked<T>(value: T, scale_with: T, offset_by: T) -> Option<T>
+where
+    T: Copy + 'static + CheckedAdd<Output = T> + CheckedMul<Output = T>,
+{
+    value
+        .checked_mul(&scale_with)
+        .and_then(|f| f.checked_add(&offset_by))
+}
+
+pub trait Scale
+where
+    Self: Sized,
+{
+    /// scales with `(self - offset_by) / scale_with`. Overflows produce `None`.
+    fn scale(self, scale_with: Self, offset_by: Self) -> Option<Self>;
+}
+
+macro_rules! impl_scale_conv {
```

Comment:
can't you implement this any differently? I mean, all the Pixel types should implement the add, mul and div traits so isn't it possible to implement this generically without macros?

Reply:
i can write it for each type...
i use checked_div, checked_mul ... for the int types since the type can overflow. And i can't impl the trait for a generic since all types implement + - * / and i can not override it (specialization). I also can not impl it in a generic way since there is an error that there might be a checked_... impl for f32, f64 later...